crates/nu-parser/src/parse_literals.rs

```diff
@@ -0,0 +1,573 @@
+use crate::{lex::lex, parser::garbage, TokenContents};
+use itertools::Itertools;
+use log::trace;
+use nu_protocol::{ast::*, engine::StateWorkingSet, ParseError, Span, Spanned, Type};
+use std::{borrow::Cow, num::ParseIntError, str};
+
+/// Strip underscores for numeric literals
+fn strip_underscores(token: &[u8]) -> Cow<[u8]> {
+    if let Some(first_underscore) = token.iter().position(|b| *b == b'_') {
+        // We assume that we parse numbers (with reasonable len) and only shortly hold the
+        // allocation, thus allocate the same size for the owned case.
+        let mut owned = Vec::with_capacity(token.len());
+        owned.extend_from_slice(&token[..first_underscore]);
+        owned.extend(token[first_underscore + 1..].iter().filter(|b| **b != b'_'));
+
+        Cow::Owned(owned)
+    } else {
+        Cow::Borrowed(token)
+    }
+}
+
+/// Failed to reinterpet bytes as an ASCII string slice
+#[derive(Debug, PartialEq)]
+struct AsciiError;
+
+/// Alternative to `str::from_utf8` if you only care about ASCII characters
+///
+/// This also allows to shortcircuit following checks if you only expect ASCII characters.
+fn parse_ascii_as_str(maybe_ascii: &[u8]) -> Result<&str, AsciiError> {
```

Comment:
`[u8]::is_ascii` does some SIMD in standard registers if possible, haven't compared on the assembly with `from_utf8` if it goes the same happy path.

As we have to perform one transform to `str` anyways I was thinking that this can shortcircuit back to parsing a string slightly faster as all the relevant numeric literals are ASCII only.